README.pod
=pod

=encoding UTF-8

=head1 NAME

Net::MAC::Vendor - Look up the vendor for a MAC

=head1 VERSION

version 1.265

=head1 SYNOPSIS

	use Net::MAC::Vendor;

	my $mac = "00:0d:93:29:f6:c2";

	my $array = Net::MAC::Vendor::lookup( $mac );

You can also run this as a script with as many arguments as you
like. The module realizes it is a script, looks up the information
for each MAC, and outputs it.

	perl Net/MAC/Vendor.pm 00:0d:93:29:f6:c2 00:0d:93:29:f6:c5

=head1 DESCRIPTION

The Institute of Electrical and Electronics Engineers (IEEE) assigns
an Organizational Unique Identifier (OUI) to manufacturers of network
interfaces. Each interface has a Media Access Control (MAC) address
of six bytes. The first three bytes are the OUI.

This module allows you to take a MAC address and turn it into the OUI
and vendor information. You can, for instance, scan a network,
collect MAC addresses, and turn those addresses into vendors. With
vendor information, you can often guess at what what you are looking
at (I<e.g.> an Apple product).

You can use this as a module as its individual functions, or call it
as a script with a list of MAC addresses as arguments. The module can
figure it out.

The IEEE moves the location of its OUI file. If they do that again, you
can set the C<NET_MAC_VENDOR_OUI_URL> environment variable to get the new
URL without updating the code.

Here are some of the old URLs, which also flip-flop schemes:

	http://standards.ieee.org/regauth/oui/oui.txt
	https://standards.ieee.org/regauth/oui/oui.txt
	http://standards-oui.ieee.org/oui.txt
	http://standards-oui.ieee.org/oui/oui.txt

There are older copies of the OUI file in the GitHub repository.

These files are large (about 4MB), so you might want to cache a copy.

A different source of information is linuxnet.ca that publishes sanitized
and compressed versions of the list, such as:

        http://linuxnet.ca/ieee/oui.txt.bz2

The module can read and decompress compressed versions (as long as the url
reflects the compression type in the filename as the linuxnet.ca links do).

=head2 Functions

=over 4

=item run( @macs )

If I call this module as a script, this class method automatically
runs. It takes the MAC addresses and prints the registered vendor
information for each address. I can pass it a list of MAC addresses
and run() processes each one of them. It prints out what it
discovers.

This method does try to use a cache of OUI to cut down on the
times it has to access the network. If the cache is fully
loaded (perhaps using C<load_cache>), it may not even use the
network at all.

=item ua

Return the Mojo::UserAgent object used to fetch resources.

=item lookup( MAC )

Given the MAC address, return an anonymous array with the vendor
information. The first element is the vendor name, and the remaining
elements are the address lines. Different records may have different
numbers of lines, although the first two should be consistent.

This makes a direct request to the IEEE website for that OUI to return
the information for that vendor.

The C<normalize_mac()> function explains the possible formats
for MAC.

=item normalize_mac( MAC )

Takes a MAC address and turns it into the form I need to
send to the IEEE lookup, which is the first six bytes in hex
separated by hyphens. For instance, 00:0d:93:29:f6:c2 turns
into 00-0D-93.

The input string can be a separated by colons or hyphens. They
can omit leading 0's (which might make things look odd). We
only need the first three bytes

	00:0d:93:29:f6:c2   # usual form

	00-0d-93-29-f6-c2   # with hyphens

	00:0d:93            # first three bytes

	0:d:93              # missing leading zero

	:d:93               # missing all leading zeros

The input string can also be a blessed L<NetAddr::MAC> object.

=item fetch_oui( MAC )

Looks up the OUI information on the IEEE website, or uses a cached
version of it. Pass it the result of C<normalize_mac()> and you
should be fine.

The C<normalize_mac()> function explains the possible formats for
MAC.

To avoid multiple calls on the network, use C<load_cache> to preload
the entire OUI space into an in-memory cache. This can take a long
time over a slow network, though; the file is about 60,000 lines.

Also, the IEEE website has been flaky lately, so loading the cache is
better. This distribution comes with several versions of the complete
OUI data file.

=item fetch_oui_from_custom( MAC, [ URL ] )

Looks up the OUI information from the specified URL or the URL set
in the C<NET_MAC_VENDOR_OUI_SOURCE> environment variable.

The C<normalize_mac()> function explains the possible formats for
MAC.

=item fetch_oui_from_ieee( MAC )

Looks up the OUI information on the IEEE website. Pass it the result
of C<normalize_mac()> and you should be fine.

The C<normalize_mac()> function explains the possible formats for
MAC.

=item fetch_oui_from_cache( MAC )

Looks up the OUI information in the cached OUI information (see
C<load_cache>).

The C<normalize_mac()> function explains the possible formats for
MAC.

To avoid multiple calls on the network, use C<load_cache> to preload
the entire OUI space into an in-memory cache.

If it doesn't find the MAC in the cache, it returns nothing.

=item extract_oui_from_html( HTML, OUI )

Gets rid of the HTML around the OUI information. It may still be
ugly. The HTML is the search results page of the IEEE ouisearch
lookup.

Returns false if it could not extract the information. This could
mean unexpected input or a change in format.

=item parse_oui( STRING )

Takes a string that looks like this:

	00-03-93   (hex)            Apple Computer, Inc.
	000393     (base 16)        Apple Computer, Inc.
								20650 Valley Green Dr.
								Cupertino CA 95014
								UNITED STATES

and turns it into an array of lines. It discards the first
line, strips the leading information from the second line,
and strips the leading whitespace from all of the lines.

With no arguments, it returns an empty anonymous array.

=item oui_url

=item oui_urls

Returns the URLs of the oui.txt resource. The IEEE likes to move this
around. These are the default URL that C<load_cache> will use, but you
can also supply your own with the C<NET_MAC_VENDOR_OUI_URL> environment
variable.

=item load_cache( [ SOURCE[, DEST ] ] )

Downloads the current list of all OUIs in SOURCE, parses it with
C<parse_oui()>, and stores it in the cache. The C<fetch_oui()> will
use this cache if it exists.

By default, this uses the URL from C<oui_url>, but given an argument,
it tries to use that.

If the url indicates that the data is compressed, the response content
is decompressed before being stored.

If C<load_cache> cannot load the data, it issues a warning and returns
nothing.

This previously used DBM::Deep if it was installed, but that was much
too slow. Instead, if you want persistence, you can play with
C<$Net::MAC::Vendor::Cached> yourself.

If you want to store the data fetched for later use, add a destination
filename to the request. To fetch from the default location and store,
specify C<undef> as source.

=back

=head1 Caching

Eventually I want people to write their own caching classes so I've
created some class methods for this.

=over 4

=item add_to_cache( OUI, PARSED_DATA )

Add to the cache. This is mostly in place for a future expansion to
full objects so you can override this in a subclass.

=item get_from_cache( OUI )

Get from the cache. This is mostly in place for a future expansion to
full objects so you can override this in a subclass.

=item get_cache_hash()

Get the hash the built-in cache uses. You should only use this if you
were using the old C<$Cached> package variable.

=back

1;
__END__

=head1 SEE ALSO

L<Net::MacMap>

=head1 SUPPORT

Bugs may be submitted through L<the RT bug tracker|https://rt.cpan.org/Public/Dist/Display.html?Name=Net-MAC-Vendor>
(or L<bug-Net-MAC-Vendor@rt.cpan.org|mailto:bug-Net-MAC-Vendor@rt.cpan.org>).

I am also usually active on irc, as 'ether' at C<irc.perl.org>.

=head1 AUTHOR

brian d foy <bdfoy@cpan.org>

=head1 CONTRIBUTORS

=for stopwords brian d foy Karen Etheridge Frank Maas openstrike Dean Hamstead

=over 4

=item *

brian d foy <brian.d.foy@gmail.com>

=item *

Karen Etheridge <ether@cpan.org>

=item *

Frank Maas <maas.frank@gmail.com>

=item *

openstrike <git@openstrike.co.uk>

=item *

Dean Hamstead <dean@fragfest.com.au>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2004 by brian d foy.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


Changes
Revision history for Perl module Net::MAC::Vendor

{{$NEXT}}
        * Fix issue where load_cache causes problems with <> after.

1.265     2019-04-12 23:55:27Z
        * properly remove Compress::* from list of mandatory prereqs

1.264     2019-04-12 19:41:33Z
	* Net::SSLeay added to runtime prerequisites.
	* Allow $mac parameter to be a NetAddr::MAC object

1.263     2019-03-26 22:26:49Z
	* Compress::Bzip2 and Compress::Zlib are now optional prerequisites if
	PUREPERL_ONLY=1 is specified as a Makefile.PL argument (or e.g. via `cpanm --pureperl`,
	or PERL_MM_OPT=PUREPERL_ONLY=1 in the environment).
	* fix warning from use of deprecated Mojo interface
	* convert packaging to Dist::Zilla, now released by ether

1.262 2017-08-10T06:45:18Z
	* Fix docs for load_cache (GitHub #23, Matt W.). Mojo::UserAgent
	doesn't handle file://

1.261 2017-07-18T18:54:10Z
	* Pete Houston (openstrike) improved fetch_oui_from_custom

1.260_03 2017-06-17T07:48:56Z
	* Try harder not to blow up tests when the network is not available

1.260_02 2017-06-09T09:00:20Z
	* Fix an errant get() call (GitHub #19 - openstrike)

1.260_01 2016-11-14T16:07:53Z
	* Fix issue with cache loading (Github #11 https://github.com/briandfoy/net-mac-vendor/issues/11)

1.26 2015-11-23T04:38:00Z
	* Bump to stable release

1.25_01 2015-11-18T04:55:45Z
	* If the tests can't connect to the fragile IEEE service,
	don't freak out.

1.25 2015-09-08T08:25:40Z
	* Bump to user version
	* CPAN Testers is down, so I hope this works.

1.24_04 2015-09-05T06:24:15Z
	* Migrate to Mojolicious for network stuff
	* Start an abstraction of the cache so others can override it
	* load_cache no longer falls back. If your source isn't there,
	it fails.

1.24_02 2015-08-08T07:51:40Z
	* Update for new IEEE URLs. Again.

1.24_01   2015-01-11 07:03:57Z
	* Fix RT #101011: Line endings are now CRLF

1.23 - 2014-11-13
	* IEEE now uses https (fixed by Frank071)
	* Cache in a file instead of a db (added by Frank071)

1.21 - 2014-01-03
	* Get rid of MYMETA

1.20 - 2013-10-03
	* Bump to release version

1.19_03 - 2013-10-01
	* Fix a test for a change in the OUI data: RT #89110

1.19_02 - 2012-11-20
	* Remove bad line from MANIFEST
	* Add encodingto Pod

1.1901 - 2012-09-03
	* Get rid of DBM::Deep. Do your own persistence. This is a
	big speedup.
	* Fix some odd parsing cases with PRIVATE records
	(RT 79369)

1.18 - 2007-10-27
	* distro cleanups after moving from CVS to SVN

1.17 - 2007-01-09
	* updated copyright and license info
	* no code changes, so no need to upgrade

1.16 - 2006-11-17
	* Fixed case-typo in docs s/Mac/MAC/. That's important to some
	operating systems. :)
	* No code updates.

1.15 - 2006-06-25
	* Fixed up missing dependency, error in fetch_oui.t that
	should have been a conditional test

1.13 - 2006-06-23
	* Improved caching that uses DBM::Deep if available
	* Added fetch_oui_from_cache, so you can avoid the network
	if you really don't want to use it
	* Added fetch_oui_from_ieee, if you only want to fetch it
	from the network (and not from cache)
	* fetch_oui now comprises those two functions, trying the
	cache first

1.11 - 2006-06-16
	* Load any oui.txt source that you like (another URL or local
	file://)

1.01 - 2006-05-25
	* The last release forgot to bump the version number. There's no
	need to upgrade since this only fixes a missing MANIFEST.

0.24 - 2006-05-25
	* MANIFEST was missing from diff. I've fixed that.

0.24 - 2006-05-20
	* Minor distro updates, kwalitee cleanups, and typo squashing.
	No need to upgrade.

0.23 - 2006-04-06
	* Added Changes to MANIFEST so you can see this file.
	* There's no need to upgrade.

0.22 - 2005-03-11
	* Added POD coverage tests and fixed problems it found
	* No code updates, so no need to upgrade

0.21 - 2004-09-07
	* added README to dist
	* no need to upgrade

0.1 - 2004-09-01
	* first pass: there is still a lot of work to do and a lot of
	tests to write

0.1 - 2004-08-31
	* first pass, a bit of a monster that grew out of a simple
	script.


scratch/mojo-test.pl
#!/Users/brian/bin/perls/perl5.14.2
use v5.10.1;
use utf8;
use strict;
use warnings;

use Mojo::UserAgent;

my $ua = Mojo::UserAgent->new->max_redirects(3);

my $url = 'https://services13.ieee.org/RST/standards-ra-web/rest/assignments/download/?registry=MA-L&format=html&text=00-0D-93';

my $tx = $ua->get( $url );

if( $tx->error ) {
	say "Failed fetching [$url]: " . $tx->res->code;
	say $tx->res->to_string;
	exit;
	}

my $html = $tx->res->body;
say $html;



Makefile.PL
package Net::MAC::Vendor;
use strict;
use warnings;

=encoding utf8

=head1 The build file for Net::MAC::Vendor

This build file is a modulino; it works as both a build script and
a module.

To build the distribution, run this file normally:

	% perl Makefile.PL

But, it's more interesting than that. You can load it with C<require>
and call C<arguments> to get the data structure it passes to
C<WriteMakefile>:

	my $package = require '/path/to/Makefile.PL';
	my $arguments = $package->arguments;

Note that C<require>-ing a file makes an entry in C<%INC> for exactly
that name. If you try to C<require> another file with the same name,
even from a different path, C<require> thinks it has already loaded
the file. As such, I recommend you always require the full path to the
file.

The return value of the C<require> is a package name (in this case,
the name of the main module). Use that to call the C<arguments> method.

Even if this distribution needs a higher version of Perl, this bit
only needs v5.8. You can play with the data structure with a primitive
Perl.

=cut

use File::Spec::Functions qw(catfile);

my $module    = __PACKAGE__;
( my $dist = $module ) =~ s/::/-/g;

my $github    = 'https://github.com/briandfoy/net-mac-vendor';
my $main_file = catfile( 'lib', split /::/, "$module.pm" );

my %WriteMakefile = (
	'MIN_PERL_VERSION' => '5.010',

	'NAME'          => $module,
	'ABSTRACT_FROM' => $main_file,
	'VERSION_FROM'  => $main_file,
	'LICENSE'       => 'artistic2',
	'AUTHOR'        => 'brian d foy <bdfoy@cpan.org>',

	'CONFIGURE_REQUIRES' => {
		'ExtUtils::MakeMaker'   => '6.64',
		'File::Spec::Functions' => '0',
		},

	'BUILD_REQUIRES' => {
		},

	'TEST_REQUIRES' => {
		'Test::More' => '0.98',
		},

	'PREREQ_PM'     => {
		'File::Temp'       => '0',
		'Mojo::URL'        => '0',
		'Mojo::UserAgent'  => '0',
		},

	'META_MERGE' => {
		dynamic_config => 1,
		'meta-spec' => { version => 2 },
		resources => {
			repository => {
				type => 'git',
				url  => "$github.git",
				web  => $github,
				},
			bugtracker => {
				web    => "$github/issues",
				},
			homepage => $github,
			},
		prereqs => {
			runtime => {
				recommends => {
					'Compress::Bzip2'  => '0',
					'Compress::Zlib'   => '0',
					},
				},
			},
		},

	clean  => { FILES    => qq|$dist-*| },
	);

sub arguments { \%WriteMakefile }

do_it() unless caller;
sub do_it {
	require File::Spec;
	my $MM ='ExtUtils::MakeMaker';
	my $MM_version =
		eval{ "$MM " . $WriteMakefile{'CONFIGURE_REQUIRES'}{'ExtUtils::MakeMaker'} }
			||
		"$MM 6.64";
	eval "use $MM_version; 1;" or die "Could not load $MM_version: $@";
	eval "use Test::Manifest 1.21"
		if -e File::Spec->catfile( qw(t test_manifest) );

	if (not parse_args()->{PUREPERL_ONLY}) {
		$WriteMakefile{PREREQ_PM}{'Compress::Bzip2'} = 0;
		$WriteMakefile{PREREQ_PM}{'Compress::Zlib'} = 0;
	}

	my $arguments = arguments();
	my $minimum_perl = $arguments->{MIN_PERL_VERSION} || '5.008';
	eval "require $minimum_perl;" or die $@;

	WriteMakefile( %$arguments );
	}

sub parse_args {
  # copied from EUMM
  require ExtUtils::MakeMaker;
  require Text::ParseWords;
  ExtUtils::MakeMaker::parse_args(
    my $tmp = {},
    Text::ParseWords::shellwords($ENV{PERL_MM_OPT} || ''),
    @ARGV,
  );
  return $tmp->{ARGS} || {};
}

no warnings;
__PACKAGE__;


lib/Net/MAC/Vendor.pm
package Net::MAC::Vendor;
# ABSTRACT: Look up the vendor for a MAC

use strict;
use warnings;
use 5.010;

use Net::SSLeay;

=head1 SYNOPSIS

	use Net::MAC::Vendor;

	my $mac = "00:0d:93:29:f6:c2";

	my $array = Net::MAC::Vendor::lookup( $mac );

You can also run this as a script with as many arguments as you
like. The module realizes it is a script, looks up the information
for each MAC, and outputs it.

	perl Net/MAC/Vendor.pm 00:0d:93:29:f6:c2 00:0d:93:29:f6:c5

=head1 DESCRIPTION

The Institute of Electrical and Electronics Engineers (IEEE) assigns
an Organizational Unique Identifier (OUI) to manufacturers of network
interfaces. Each interface has a Media Access Control (MAC) address
of six bytes. The first three bytes are the OUI.

This module allows you to take a MAC address and turn it into the OUI
and vendor information. You can, for instance, scan a network,
collect MAC addresses, and turn those addresses into vendors. With
vendor information, you can often guess at what what you are looking
at (I<e.g.> an Apple product).

You can use this as a module as its individual functions, or call it
as a script with a list of MAC addresses as arguments. The module can
figure it out.

The IEEE moves the location of its OUI file. If they do that again, you
can set the C<NET_MAC_VENDOR_OUI_URL> environment variable to get the new
URL without updating the code.

Here are some of the old URLs, which also flip-flop schemes:

	http://standards.ieee.org/regauth/oui/oui.txt
	https://standards.ieee.org/regauth/oui/oui.txt
	http://standards-oui.ieee.org/oui.txt
	http://standards-oui.ieee.org/oui/oui.txt

There are older copies of the OUI file in the GitHub repository.

These files are large (about 4MB), so you might want to cache a copy.

A different source of information is linuxnet.ca that publishes sanitized
and compressed versions of the list, such as:

        http://linuxnet.ca/ieee/oui.txt.bz2

The module can read and decompress compressed versions (as long as the url
reflects the compression type in the filename as the linuxnet.ca links do).

=head2 Functions

=over 4

=cut

use Exporter qw(import);

__PACKAGE__->run( @ARGV ) unless caller;

use Carp ();
use Mojo::URL;
use Mojo::UserAgent;

our $VERSION = '1.266';

=item run( @macs )

If I call this module as a script, this class method automatically
runs. It takes the MAC addresses and prints the registered vendor
information for each address. I can pass it a list of MAC addresses
and run() processes each one of them. It prints out what it
discovers.

This method does try to use a cache of OUI to cut down on the
times it has to access the network. If the cache is fully
loaded (perhaps using C<load_cache>), it may not even use the
network at all.

=cut

sub run {
	my $class = shift;

	foreach my $arg ( @_ ) {
		my $lines = lookup( $arg );
		return unless defined $lines;

		unshift @$lines, $arg;

		print join "\n", @$lines, '';
		}

	return 1;
	}

=item ua

Return the Mojo::UserAgent object used to fetch resources.

=cut

sub ua {
	state $ua = Mojo::UserAgent->new->max_redirects(3);
	$ua;
	}

=item lookup( MAC )

Given the MAC address, return an anonymous array with the vendor
information. The first element is the vendor name, and the remaining
elements are the address lines. Different records may have different
numbers of lines, although the first two should be consistent.

This makes a direct request to the IEEE website for that OUI to return
the information for that vendor.

The C<normalize_mac()> function explains the possible formats
for MAC.

=cut

sub lookup {
	my $mac   = shift;

	   $mac   = normalize_mac( $mac );
	my $lines = fetch_oui( $mac );

	return $lines;
	}

=item normalize_mac( MAC )

Takes a MAC address and turns it into the form I need to
send to the IEEE lookup, which is the first six bytes in hex
separated by hyphens. For instance, 00:0d:93:29:f6:c2 turns
into 00-0D-93.

The input string can be a separated by colons or hyphens. They
can omit leading 0's (which might make things look odd). We
only need the first three bytes

	00:0d:93:29:f6:c2   # usual form

	00-0d-93-29-f6-c2   # with hyphens

	00:0d:93            # first three bytes

	0:d:93              # missing leading zero

	:d:93               # missing all leading zeros

The input string can also be a blessed L<NetAddr::MAC> object.

=cut

sub normalize_mac {
	no warnings 'uninitialized';

	my $input = shift;

	return uc($input->as_microsoft)
	    if ref $input eq 'NetAddr::MAC';

	$input = uc $input;

	do {
		Carp::carp "Could not normalize MAC [$input]";
		return
		} if $input =~ m/[^0-9a-f:-]/i;

	my @bytes =
		grep { /^[0-9A-F]{2}$/ }
		map { sprintf "%02X", hex }
		grep { defined }
		( split /[:-]/, $input )[0..2];

	do {
		Carp::carp "Could not normalize MAC [$input]";
		return
		} unless @bytes == 3;

	my $mac = join "-", @bytes;

	return $mac;
	}

=item fetch_oui( MAC )

Looks up the OUI information on the IEEE website, or uses a cached
version of it. Pass it the result of C<normalize_mac()> and you
should be fine.

The C<normalize_mac()> function explains the possible formats for
MAC.

To avoid multiple calls on the network, use C<load_cache> to preload
the entire OUI space into an in-memory cache. This can take a long
time over a slow network, though; the file is about 60,000 lines.

Also, the IEEE website has been flaky lately, so loading the cache is
better. This distribution comes with several versions of the complete
OUI data file.

=cut

sub fetch_oui {
	# fetch_oui_from_custom( $_[0] )    ||
		fetch_oui_from_cache( $_[0] ) ||
		fetch_oui_from_ieee( $_[0] );
	}

=item fetch_oui_from_custom( MAC, [ URL ] )

Looks up the OUI information from the specified URL or the URL set
in the C<NET_MAC_VENDOR_OUI_SOURCE> environment variable.

The C<normalize_mac()> function explains the possible formats for
MAC.

=cut

sub fetch_oui_from_custom {
	my $mac = normalize_mac( shift );
	my $url = shift // $ENV{NET_MAC_VENDOR_OUI_SOURCE};

	return unless defined $url;

	my $html = __PACKAGE__->_fetch_oui_from_url( $url );
	unless( defined $html ) {
		Carp::carp "Could not fetch data from the IEEE!";
		return;
		}

	parse_oui(
		extract_oui_from_html( $html, $mac )
		);
	}

=item fetch_oui_from_ieee( MAC )

Looks up the OUI information on the IEEE website. Pass it the result
of C<normalize_mac()> and you should be fine.

The C<normalize_mac()> function explains the possible formats for
MAC.

=cut

sub _search_url_base {
# https://services13.ieee.org/RST/standards-ra-web/rest/assignments/download/?registry=MA-L&format=html&text=00-0D-93
	state $url = Mojo::URL->new(
		'https://services13.ieee.org/RST/standards-ra-web/rest/assignments/download/?registry=MA-L&format=html'
		);

	$url;
	}

sub _search_url {
	my( $class, $mac ) = @_;
	my $url = $class->_search_url_base->clone;
	$url->query->merge( text => $mac );
	$url;
	}

sub fetch_oui_from_ieee {
	my $mac = normalize_mac( shift );

	my @urls = __PACKAGE__->_search_url( $mac );

	my $html;
	URL: foreach my $url ( @urls ) {
		$html = __PACKAGE__->_fetch_oui_from_url( $url );
		next URL unless defined $html;
		last;
		}

	unless( defined $html ) {
		Carp::carp "Could not fetch data from the IEEE!";
		return;
		}

	parse_oui(
		extract_oui_from_html( $html, $mac )
		);
	}

sub _fetch_oui_from_url {
	state $min_ssl = 0x10_00_00_00;
	my( $class, $url ) = @_;
	my $tries = 0;

	my $ssl_version =  Net::SSLeay::SSLeay();
	my $ssl_version_string = Net::SSLeay::SSLeay_version();

	if( $ssl_version < $min_ssl ) {
		Carp::carp "Fetching OUI might fail with older OpenSSLs. You have [$ssl_version_string] and may need 1.x";
		}

	return unless defined $url;

	TRY: {
		my $tx = __PACKAGE__->ua->get( $url );
		if( $tx->error ) {
			if( $tries > 3 ) {
				my $error  = $tx->error;
				my @messages = (
					"Failed fetching [$url] HTTP status [$error->{code}]",
					"message [$error->{message}]"
					);
				push @messages, "You may need to upgrade OpenSSL to 1.x. You have [$ssl_version_string]"
					if $ssl_version < $min_ssl;

				Carp::carp join "\n", @messages;
				return;
				}

			$tries++;
			sleep 1 * $tries;
			redo TRY;
			}

		my $html = $tx->res->body;
		unless( defined $html ) {
			Carp::carp "No content in response for [$url]!";
			return;
			}

		return $html;
		}
	}

=item fetch_oui_from_cache( MAC )

Looks up the OUI information in the cached OUI information (see
C<load_cache>).

The C<normalize_mac()> function explains the possible formats for
MAC.

To avoid multiple calls on the network, use C<load_cache> to preload
the entire OUI space into an in-memory cache.

If it doesn't find the MAC in the cache, it returns nothing.

=cut

sub fetch_oui_from_cache {
	my $mac = normalize_mac( shift );

	__PACKAGE__->get_from_cache( $mac );
	}

=item extract_oui_from_html( HTML, OUI )

Gets rid of the HTML around the OUI information. It may still be
ugly. The HTML is the search results page of the IEEE ouisearch
lookup.

Returns false if it could not extract the information. This could
mean unexpected input or a change in format.

=cut

sub extract_oui_from_html {
	my $html = shift;
	my $lookup_mac = normalize_mac( shift );

	my( $record ) = $html =~ m|<pre>(<b>$lookup_mac</b>.*?)</pre>|is;
	$record =~ s|</?b>||g;

	return unless defined $record;
	return $record;
	}

=item parse_oui( STRING )

Takes a string that looks like this:

	00-03-93   (hex)            Apple Computer, Inc.
	000393     (base 16)        Apple Computer, Inc.
								20650 Valley Green Dr.
								Cupertino CA 95014
								UNITED STATES

and turns it into an array of lines. It discards the first
line, strips the leading information from the second line,
and strips the leading whitespace from all of the lines.

With no arguments, it returns an empty anonymous array.

=cut

sub parse_oui {
	my $oui = shift;
	return [] unless $oui;
	$oui =~ s|</?b>||g;
	my @lines = map { s/^\s+//; $_ ? $_ : () } split /\s*$/m, $oui;
	chomp @lines;
	splice @lines, 1, 1, (); # should have documented this!

	$lines[0] =~ s/\S+\s+\S+\s+//;
	return \@lines;
	}

=item oui_url

=item oui_urls

Returns the URLs of the oui.txt resource. The IEEE likes to move this
around. These are the default URL that C<load_cache> will use, but you
can also supply your own with the C<NET_MAC_VENDOR_OUI_URL> environment
variable.

=cut

sub oui_url { (grep { /\Ahttp:/ } &oui_urls)[0] }

sub oui_urls {
	my @urls = 'http://standards-oui.ieee.org/oui.txt';

	unshift @urls, $ENV{NET_MAC_VENDOR_OUI_URL}
		if defined $ENV{NET_MAC_VENDOR_OUI_URL};

	@urls;
	}

=item load_cache( [ SOURCE[, DEST ] ] )

Downloads the current list of all OUIs in SOURCE, parses it with
C<parse_oui()>, and stores it in the cache. The C<fetch_oui()> will
use this cache if it exists.

By default, this uses the URL from C<oui_url>, but given an argument,
it tries to use that.

If the url indicates that the data is compressed, the response content
is decompressed before being stored.

If C<load_cache> cannot load the data, it issues a warning and returns
nothing.

This previously used DBM::Deep if it was installed, but that was much
too slow. Instead, if you want persistence, you can play with
C<$Net::MAC::Vendor::Cached> yourself.

If you want to store the data fetched for later use, add a destination
filename to the request. To fetch from the default location and store,
specify C<undef> as source.

=cut

sub load_cache {
	my( $source, $dest ) = @_;

	my $data = do {;
		if( defined $source ) {
			unless( -e $source ) {
				Carp::carp "Net::Mac::Vendor cache source [$source] does not exist";
				return;
				}

			do { local( *ARGV, $/ ); @ARGV = $source; <> }
			}
		else {
			#say time . " Fetching URL";
			my $url = oui_url();
			my $tx = __PACKAGE__->ua->get( $url );
			#say time . " Fetched URL";
			#say "size is " . $tx->res->headers->header( 'content-length' );
			($url =~ /\.bz2/) ? _bunzip($tx->res->body) :
			($url =~ /\.gz/)  ? _gunzip($tx->res->body) :
			                    $tx->res->body;
			}
		};

	if( defined $dest ) {
		if( open my $fh, '>:utf8', $dest ) {
			print { $fh } $data;
			close $fh;
			}
		else { # notify on error, but continue
			Carp::carp "Could not write to '$dest': $!";
			}
		}


	# The PRIVATE entries fill in a template with no
	# company name or address, but the whitespace is
	# still there. We need to split on a newline
	# followed by some potentially horizontal whitespace
	# and another newline
	my $CRLF = qr/(?:\r?\n)/;
	my @entries = split /[\t ]* $CRLF [\t ]* $CRLF/x, $data;
	shift @entries;

	my $count = '';
	foreach my $entry ( @entries ) {
		$entry =~ s/^\s+//;
		my $oui = substr $entry, 0, 8;
		__PACKAGE__->add_to_cache( $oui, parse_oui( $entry ) );
		}

	return 1;
	}

sub _bunzip {
	my $content = shift;
	if (eval { +require Compress::Bzip2; 1 }) {
		return Compress::Bzip2::memBunzip($content);
		}
	else {
		require File::Temp;
		my ($tempfh, $tempfilename) = File::Temp::tempfile( UNLINK => 1 );
		binmode $tempfh, ':raw';
		print $tempfh $content;
		close $tempfh;

		open my $unzipfh, "bunzip2 --stdout $tempfilename |"
			or die "cannot pipe to bunzip2: $!";
		local $/;
		return <$unzipfh>;
		}
	}

sub _gunzip {
	my $content = shift;
	if (eval { +require Compress::Zlib; 1 }) {
		return Compress::Zlib::memGunzip($content);
		}
	else {
		require File::Temp;
		my ($tempfh, $tempfilename) = File::Temp::tempfile( UNLINK => 1 );
		binmode $tempfh, ':raw';
		print $tempfh $content;
		close $tempfh;

		open my $unzipfh, "gunzip --stdout $tempfilename |"
			or die "cannot pipe to gunzip: $!";
		local $/;
		return <$unzipfh>;
		}
	}

=back

=head1 Caching

Eventually I want people to write their own caching classes so I've
created some class methods for this.

=over 4

=cut

BEGIN {
my $Cached = {};

=item add_to_cache( OUI, PARSED_DATA )

Add to the cache. This is mostly in place for a future expansion to
full objects so you can override this in a subclass.

=cut

sub add_to_cache {
	my( $class, $oui, $parsed ) = @_;

	$Cached->{ $oui } = $parsed;
	}

=item get_from_cache( OUI )

Get from the cache. This is mostly in place for a future expansion to
full objects so you can override this in a subclass.

=cut

sub get_from_cache {
	my( $class, $oui ) = @_;

	$Cached->{ $oui };
	}

=item get_cache_hash()

Get the hash the built-in cache uses. You should only use this if you
were using the old C<$Cached> package variable.

=cut

sub get_cache_hash { $Cached }
}

=back

1;
__END__

=head1 SEE ALSO

L<Net::MacMap>

=cut
